plugins/gstreamer/src/GStreamerVideoSource.cpp

```diff
@@ -0,0 +1,299 @@
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+ * associated documentation files (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:  The above copyright notice and this
+ * permission notice (including the next paragraph) shall be included in all copies or substantial
+ * portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <charconv>
+#include <cstdlib>
+
+#include <gst/gst.h>
+#include <gst/app/gstappsink.h>
+
+#include "GStreamerVideoSource.h"
+
+namespace neurala
+{
+namespace
+{
+class Sample
+{
+	GstSample* sample;
+	GstBuffer* buffer;
+	GstMapInfo map;
+
+public:
+	explicit Sample(GstSample* sample) noexcept
+	 : sample(sample), buffer(gst_sample_get_buffer(sample))
+	{
+		gst_buffer_map(buffer, &map, GST_MAP_READ);
+	}
+
+	~Sample() noexcept
+	{
+		gst_buffer_unmap(buffer, &map);
+		gst_sample_unref(sample);
+	}
+
+	dto::ImageView imageView(std::size_t width, std::size_t height) const noexcept
+	{
+		return dto::ImageView(dto::ImageMetadata("uint8", width, height, "RGB", "interleaved", "topLeft"), map.data);
+	}
+
+	Sample(const Sample&) = delete;
+	Sample(Sample&&) = delete;
+
+	Sample& operator=(const Sample&) = delete;
+	Sample& operator=(Sample&&) = delete;
+};
+}
+
+struct GStreamerVideoSource::Implementation
+{
+	GstElement* pipeline;
+	GstElement* userPipeline;
+	GstElement* sink;
+
+	std::unique_ptr<Sample> sample;
+};
+
+void*
+GStreamerVideoSource::create(PluginArguments&, PluginErrorCallback&)
+{
+	if(!gst_is_initialized())
+	{
+		gst_init(0, NULL);
+	}
+	return new GStreamerVideoSource("PLACEHOLDER");
+}
+
+void
+GStreamerVideoSource::destroy(void* p)
+{
+	delete static_cast<GStreamerVideoSource*>(p);
+}
+
+GStreamerVideoSource::GStreamerVideoSource(const char* name)
+ : m_implementation(std::make_unique<Implementation>()),
+   m_bufferReady(false),
+   m_frameReady(false),
+   m_endOfStream(false)
+{
+	m_implementation->pipeline = gst_pipeline_new(name);
+
+	{
+		const auto parser = gst_parse_context_new();
+		const auto userPipeline = getenv("NEURALA_GSTREAMER_PIPELINE");
+
+		if (!userPipeline)
+		{
+			m_lastError = B4BError::invalidParameter();
+			return;
+		}
+
+		m_implementation->userPipeline = gst_parse_launch(userPipeline, nullptr);
```

Comment:
Looks like you can't just add the sink to the bin: https://stackoverflow.com/a/52188591

We should just require that user always end their pipeline in an appsink name `neuralaviasink` for example.

Reply:
That's actually my current solution, now. The sink is added to the user pipeline, then the name is given to us.

Reply:
I think this is technically a more flexible solution, as the user could do nigh anything with their pipeline (like tee to our `appsink` and to whatever else they may need).